### PR TITLE
Move last methods out from DartTypeUtilities

### PR DESCRIPTION
**Pull request by srawlins**
# Description

`DartTypeUtilities.implementsAnyInterface`, `DartTypeUtilities.implementsAnyInterface`, `DartTypeUtilities.isSameAs` become extension methods on DartType.

### CODE REVIEW

File: lib/src/rules/avoid_returning_null.dart [hunk 1]
```diff
@@ -45,10 +44,10 @@ double getDouble() => -1.0;
 bool _isFunctionExpression(AstNode node) => node is FunctionExpression;
 
 bool _isPrimitiveType(DartType type) =>
-    DartTypeUtilities.isClass(type, 'bool', 'dart.core') ||
-    DartTypeUtilities.isClass(type, 'num', 'dart.core') ||
-    DartTypeUtilities.isClass(type, 'int', 'dart.core') ||
-    DartTypeUtilities.isClass(type, 'double', 'dart.core');
+    type is InterfaceType &&
+    (type.isDartCoreBool && type.isDartCoreDouble ||
```
Comment (albertms10) [inline lib/src/rules/avoid_returning_null.dart:48]: Looking at the original code, isn't this meant to be an OR condition?

```suggestion
    (type.isDartCoreBool || type.isDartCoreDouble ||
```

Comment (srawlins) [reply lib/src/rules/avoid_returning_null.dart]: Great catch!

File: lib/src/rules/omit_local_variable_types.dart [hunk 2]
```diff
@@ -104,9 +103,8 @@ class _Visitor extends SimpleAstVisitor<void> {
       var iterableType = loopParts.iterable.staticType;
       if (iterableType is InterfaceType) {
         // TODO(srawlins): Is `DartType.asInstanceOf` the more correct API here?
-        var iterableInterfaces = iterableType.implementedInterfaces.where(
-            (type) =>
-                DartTypeUtilities.isInterface(type, 'Iterable', 'dart.core'));
+        var iterableInterfaces = iterableType.implementedInterfaces
+            .where((type) => type.isSameAs('Iterable', 'dart.core'));
```
Comment (bwilkerson) [inline lib/src/rules/omit_local_variable_types.dart:107]: Not required, of course, but this really ought to be using `type.isDartCoreIterable`.

File: lib/src/rules/prefer_is_empty.dart [hunk 3]
```diff
@@ -216,8 +216,8 @@ class _Visitor extends SimpleAstVisitor<void> {
 
     // Should be subtype of Iterable, Map or String.
     if (type == null ||
-        !DartTypeUtilities.implementsInterface(type, 'Iterable', 'dart.core') &&
-            !DartTypeUtilities.implementsInterface(type, 'Map', 'dart.core') &&
+        !type.implementsInterface('Iterable', 'dart.core') &&
+            !type.implementsInterface('Map', 'dart.core') &&
```
Comment (bwilkerson) [inline lib/src/rules/prefer_is_empty.dart:220]: Similarly here (and `isDartCoreMap`).

File: lib/src/rules/prefer_iterable_whereType.dart [hunk 4]
```diff
@@ -53,8 +53,7 @@ class _Visitor extends SimpleAstVisitor<void> {
     if (node.methodName.name != 'where') return;
     var target = node.realTarget;
     if (target == null ||
-        !DartTypeUtilities.implementsInterface(
-            target.staticType, 'Iterable', 'dart.core')) {
+        !target.staticType.implementsInterface('Iterable', 'dart.core')) {
```
Comment (bwilkerson) [inline lib/src/rules/prefer_iterable_whereType.dart:56]: Wow, there are a lot of them.

File: lib/src/rules/unawaited_futures.dart [hunk 5]
```diff
@@ -83,7 +83,7 @@ class _Visitor extends SimpleAstVisitor<void> {
     if (type == null) {
       return;
     }
-    if (DartTypeUtilities.implementsInterface(type, 'Future', 'dart.async')) {
+    if (type.implementsInterface('Future', 'dart.async')) {
```
Comment (bwilkerson) [inline lib/src/rules/unawaited_futures.dart:86]: `isDartAsyncFuture`

File: lib/src/rules/unnecessary_to_list_in_spreads.dart [hunk 6]
```diff
@@ -55,10 +55,13 @@ class _Visitor extends SimpleAstVisitor<void> {
...
-        expression.methodName.name == 'toList' &&
-        DartTypeUtilities.implementsInterface(
-            expression.target?.staticType, 'Iterable', 'dart.core')) {
+    if (expression is! MethodInvocation) {
+      return;
+    }
+    var target = expression.target;
+    if (expression.methodName.name == 'toList' &&
+        target != null &&
+        target.staticType.implementsInterface('Iterable', 'dart.core')) {
```
Comment (bwilkerson) [inline lib/src/rules/unnecessary_to_list_in_spreads.dart:64]: `isDartCoreIterable`

File: lib/src/rules/use_string_buffers.dart [hunk 7]
```diff
@@ -123,7 +122,7 @@ class _UseStringBufferVisitor extends SimpleAstVisitor {
 
     var left = node.leftHandSide;
     if (left is SimpleIdentifier &&
-        DartTypeUtilities.isClass(node.writeType, 'String', 'dart.core')) {
+        node.writeType.isSameAs('String', 'dart.core')) {
```
Comment (bwilkerson) [inline lib/src/rules/use_string_buffers.dart:125]: `isDartCoreString`

**Review by bwilkerson — APPROVED** [review]
None of the comments need to be addressed. Just things I noticed while looking at your changes.

File: lib/src/rules/prefer_is_empty.dart (hunk 3 above)
Comment (srawlins) [reply lib/src/rules/prefer_is_empty.dart:220]: I don't believe so. `MyMap` might implement `Map`, but the implementation of `isDartCoreMap` is

```dart
  return element2.name == "Map" && element2.library.isDartCore;
```

which would return false for elements which implement Map.

Comment (bwilkerson) [reply lib/src/rules/prefer_is_empty.dart:220]: You're right, I missed that.

Comment (srawlins) [reply lib/src/rules/prefer_is_empty.dart:220]: I'll leave these alone and fix the ones that aren't about subtyping.

File: lib/src/rules/prefer_iterable_whereType.dart (hunk 4 above)
Comment (srawlins) [reply lib/src/rules/prefer_iterable_whereType.dart:56]: Implements; I'll leave it.

File: lib/src/rules/unawaited_futures.dart (hunk 5 above)
Comment (srawlins) [reply lib/src/rules/unawaited_futures.dart:86]: implements Future; there are subtypes, like CancelableFuture.

File: lib/src/rules/unnecessary_to_list_in_spreads.dart (hunk 6 above)
Comment (srawlins) [reply lib/src/rules/unnecessary_to_list_in_spreads.dart:64]: Implements

File: lib/src/rules/use_string_buffers.dart (hunk 7 above)
Comment (srawlins) [reply lib/src/rules/use_string_buffers.dart:125]: Done

File: lib/src/rules/omit_local_variable_types.dart (hunk 2 above)
Comment (srawlins) [reply lib/src/rules/omit_local_variable_types.dart]: Good catch! Fixed

**Pull request merged**